framework/src/org/checkerframework/common/value/RangeOrListOfValues.java

```diff
@@ -0,0 +1,116 @@
+package org.checkerframework.common.value;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.lang.model.element.AnnotationMirror;
+import org.checkerframework.common.value.qual.ArrayLen;
+import org.checkerframework.common.value.qual.ArrayLenRange;
+import org.checkerframework.common.value.qual.IntVal;
+import org.checkerframework.common.value.util.Range;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+
+/**
+ * An abstraction that can be either a range or a list of values that could come from an {@link
+ * ArrayLen} or {@link IntVal}. This abstraction reduces the number of cases that {@link
+ * org.checkerframework.common.value.ValueAnnotatedTypeFactory.ValueTreeAnnotator#handleInitializers(List,
+ * AnnotatedTypeMirror.AnnotatedArrayType)} and {@link
+ * org.checkerframework.common.value.ValueAnnotatedTypeFactory.ValueTreeAnnotator#handleDimensions(List,
+ * AnnotatedTypeMirror.AnnotatedArrayType)} must handle.
+ *
+ * <p>Tracks Ints in the list, and creates ArrayLen or ArrayLenRange annotations, because it's meant
+ * to be used to reason about ArrayLen and ArrayLenRange values.
+ */
+class RangeOrListOfValues {
+    private Range range;
+    private List<Integer> values;
+    private boolean isRange;
+
+    public RangeOrListOfValues() {
+        this.values = new ArrayList<>();
+        isRange = false;
+    }
+
+    public RangeOrListOfValues(List<Integer> values) {
+        this.values = new ArrayList<>();
+        isRange = false;
+        addAll(values);
+    }
+
+    public RangeOrListOfValues(Range range) {
+        this.range = range;
+        isRange = true;
+    }
+
+    public void add(Range otherRange) {
+        if (isRange) {
+            range = range.union(otherRange);
+        } else {
+            convertToRange();
+            add(otherRange);
+        }
+    }
+
+    /**
+     * If this is not a range, adds all members of newValues to the list. Otherwise, extends the
+     * range as appropriate based on the max and min of newValues. If adding newValues to a
+     * non-range would cause the list to become too large, converts this into a range.
```

Comment:
Please mention that `convertLongsToInts` should be used if to convert `List<Long>' to 'List<Integer>'.